thingif/src/main/java/com/kii/thingif/ThingIFAPI.java

```diff
@@ -2170,7 +2173,51 @@ public String getFirmwareVersion() throws ThingIFException {
             @NonNull Aggregation aggregation,
             @NonNull Class<S> targetStateClass,
             @NonNull Class<T> valueClass) throws ThingIFException {
-        //TODO: // FIXME: 12/21/16 implement the logic
-        return new ArrayList<>();
+        if (this.target == null) {
+            throw new IllegalStateException("Can not perform this action before onboarding");
+        }
+        if (!this.stateTypes.containsKey(groupedQuery.getAlias())) {
+            throw new UnregisteredAliasException(groupedQuery.getAlias(), false);
+        }
+
+        Gson localGson = new GsonBuilder()
+                .registerTypeAdapter(GroupedHistoryStatesQuery.class,
+                        new GroupedHistoryStatesQueryAdapter(aggregation))
+                .registerTypeAdapter(AggregatedResult.class,
+                        new AggregatedResultAdapter(targetStateClass, valueClass))
+                .create();
+
+        String path = MessageFormat.format("/thing-if/apps/{0}/targets/{1}/states/aliases/{2}/query",
+                this.app.getAppID(), this.target.getTypedID().toString(), groupedQuery.getAlias());
+        String url = Path.combine(this.app.getBaseUrl(), path);
+        Map<String, String> headers = this.newHeader();
+        JSONObject requestBody = JsonUtils.newJson(localGson.toJson(groupedQuery));
+        IoTRestRequest request = new IoTRestRequest(url, IoTRestRequest.Method.POST, headers,
+                MediaTypes.MEDIA_TYPE_TRAIT_STATE_QUERY_REQUEST, requestBody);
+
+        List<AggregatedResult<T, S>> retList = new ArrayList<>();
+
+        JSONObject responseBody;
+        try {
+            responseBody = this.restClient.sendRequest(request);
+        }catch (ConflictException e) {
+            // when thing never update its state to server, server returns this error.
+            if (e.getErrorCode() != null && e.getErrorCode().equals("STATE_HISTORY_NOT_AVAILABLE")) {
+                return retList;
+            } else {
+                throw e;
+            }
+        }
+
+        JSONArray results = responseBody.optJSONArray("groupedResults");
+        if (results != null) {
+            for (int i = 0; i < results.length(); ++i) {
+                JSONObject result = results.optJSONObject(i);
+                if (result != null) {
+                    retList.add(localGson.fromJson(result.toString(), AggregatedResult.class));
```

Comment:
make sure to fix uncheck warning caused by parsing generic for gson, fix like this: 

```java
        Type aggregatedResultType = new TypeToken<AggregatedResult<T, S>>(){}.getType();
        if (results != null) {
            for (int i = 0; i < results.length(); ++i) {
                JSONObject result = results.optJSONObject(i);
                if (result != null) {
                    AggregatedResult<T, S> aggregatedResult = 
                            localGson.fromJson(result.toString(), aggregatedResultType);
                    retList.add(aggregatedResult);
                }
            }
        }
```